src/neural-network.js

```diff
@@ -239,47 +262,103 @@ export default class NeuralNetwork {
     let learningRate = _options.learningRate || this.learningRate || options.learningRate;
     let callback = options.callback;
     let callbackPeriod = options.callbackPeriod;
-    if (!options.reinforce) {
-      let sizes = [];
-      let inputSize = data[0].input.length;
-      let outputSize = data[0].output.length;
-      let hiddenSizes = this.hiddenSizes;
-      if (!hiddenSizes) {
-        sizes.push(Math.max(3, Math.floor(inputSize / 2)));
-      } else {
-        hiddenSizes.forEach(size => {
-          sizes.push(size);
-        });
-      }
-
-      sizes.unshift(inputSize);
-      sizes.push(outputSize);
+    let endTime = Date.now() + options.trainTimeMs;
+    var res = {
+      error: 1,
+      iterations: 0
+    };
 
-      this.initialize(sizes);
+    if (this.sizes === null) {
+      let sizes = this._getSizesFromData (data);
+      this.initialize (sizes);
     }
 
-    let error = 1;
-    let i;
-    for (i = 0; i < iterations && error > errorThresh; i++) {
+    while (res.iterations < iterations && res.error > errorThresh && Date.now () > endTime) {
+      res.iterations++;
       let sum = 0;
-      for (let j = 0; j < data.length; j++) {
-        let err = this.trainPattern(data[j].input, data[j].output, learningRate);
-        sum += err;
+      for (var i = 0; i < data.length; ++i) {
+        sum += this.trainPattern (data[i].input, data[i].output, learningRate);
       }
-      error = sum / data.length;
 
-      if (log && (i % logPeriod === 0)) {
-        log('iterations:', i, 'training error:', error);
+      res.error = sum / data.length;
+
+      if (log && (res.iterations % logPeriod === 0)) {
+        log('iterations:', res.iterations, 'training error:', res.error);
       }
-      if (callback && (i % callbackPeriod === 0)) {
-        callback({ error: error, iterations: i });
+
+      if (callback && (res.iterations % callbackPeriod === 0)) {
```

Comment:
The callback is for getting periodic training results? Not the finished trained model?

Reply:
yeah that's how it is currently setup.

Reply:
Why?

Reply:
logging for the most part.  Most people like to see training advance.  I don't see this going away any time soon, as it was part of the model when the community took over.

Reply:
Isn't that what the `log` option does? It's cool though, separate conversation for the community, this maintains current functionality